internal/archive/credentials.go

```diff
@@ -0,0 +1,297 @@
+// username and password if they are specified in repoUrl.
+func lookupFor(repoUrl string, creds *credentials) (*credentialsLookup, error) {
+	u, err := url.Parse(repoUrl)
+	if err != nil {
+		return nil, err
+	}
+	host := u.Host
+	port := u.Port()
+	if port != "" {
+		// u.Hostname() would remove brackets from IPv6 address but we
+		// need it verbatim for string comparison in netrc file. This
+		// is also a bit faster because both u.Port() and u.Hostname()
+		// split u.Host into port and hostname.
+		host = u.Host[0 : len(u.Host)-len(port)-1]
+	}
+
+	lookup := credentialsLookup{
+		scheme:     u.Scheme,
+		host:       host,
+		port:       port,
+		path:       u.Path,
+		// If the input URL specifies unencrypted scheme, the scheme in
+		// machine declarations in netrc file is not optional and must
+		// also match.
+		needScheme: u.Scheme != "https" && u.Scheme != "tor+https",
+	}
+
+	if creds != nil {
+		creds.Username = u.User.Username()
+		creds.Password, _ = u.User.Password()
+	}
+
+	return &lookup, nil
+}
+
+// findCredentials searches credentials for repoUrl in configuration files in
+// directory specified by CHISEL_AUTH_DIR environment variable if it's
+// non-empty or /etc/apt/auth.conf.d.
+func findCredentials(repoUrl string) (credentials, error) {
+	credentialsDir := "/etc/apt/auth.conf.d"
+	if v := os.Getenv("CHISEL_AUTH_DIR"); v != "" {
+		credentialsDir = v
+	}
+	return findCredentialsDir(repoUrl, credentialsDir)
+}
+
+// findCredentialsDir searches for credentials for repoUrl in configuration
+// files in credentialsDir directory. If the directory does not exist, empty
+// credentials structure with nil err is returned.
+// Only files that do not begin with dot and have either no or ".conf"
+// extension are searched. The files are searched in ascending lexicographic
+// order. The first file that contains machine declaration matching repoUrl
+// ends the search. If no file contain matching machine declaration, empty
+// credentials structure with nil err is returned.
+func findCredentialsDir(repoUrl string, credentialsDir string) (creds credentials, err error) {
+	contents, err := os.ReadDir(credentialsDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			err = nil
+		} else {
+			err = fmt.Errorf("cannot open credentials directory: %w", err)
+		}
+		return
+	}
+
+	confFiles := make([]string, 0, len(contents))
+	for _, entry := range contents {
+		name := entry.Name()
+		if strings.HasPrefix(name, ".") {
+			continue
+		}
+		var info fs.FileInfo
+		info, err = entry.Info()
+		if err != nil {
+			return
+		}
+		if !info.Mode().IsRegular() {
+			continue
+		}
+		ext := filepath.Ext(name)
+		if ext == "" || ext == ".conf" {
+			confFiles = append(confFiles, name)
+		}
+	}
+	sort.Strings(confFiles)
+
+	lookup, err := lookupFor(repoUrl, &creds)
+	if err != nil {
+		return
+	}
+
+	errs := make([]error, 0, len(confFiles))
+
+	for _, file := range confFiles {
+		if !creds.Empty() {
+			break
+		}
```

Comment:
> creds can be non-empty after the lookupFor call on Ln 128 I think.

It can in case username and password are set in the URL string (`https://user:pass@example.com/foo`).

> But I feel it's better to have this check right after the lookup (after Ln 131). And, also below after the findCredentialsFile call in Ln 146.

Why? This is shorther and does the same thing. :-)

Reply:
On second thought, you are right, that with immediate check after we can skip some code. I also moved config files sort after the check. Updated in https://github.com/canonical/chisel/pull/60/commits/caabf58cf80028fed0e3c4ed52e6f163faee8d1d#diff-7e64e820d9fd9bdb86235e7058189c2ab75ebe9af60e2c9ec79db718dc5e6a75R139